fabric-rendering-v1/src/main/java/net/fabricmc/fabric/api/client/rendering/v1/WorldRenderContext.java

```diff
@@ -0,0 +1,203 @@
+		 * must be drawn directly to the frame buffer, preferably in {@link WorldRenderEvents#LAST} to avoid being
+		 * overdrawn or cleared.
+		 */
+		VertexConsumerProvider consumers();
+	}
+
+	/**
+	 * Conveys WorldRenderer state to event handlers after invocation.
+	 * Includes view frustum, which is not available then.
+	 */
+	public interface AfterFrustum {
+		Frustum frustum();
+	}
+
+	/**
+	 * Used in the block breaking callback to convey information on
+	 * ray trace hit and accept cancellation of default outline render.
+	 */
+	public interface PreBlockOutline {
+		/**
+		 * The game object currently under the crosshair target. Normally
+		 * equivalent to {@link MinecraftClient#crosshairTarget}.
+		 * Provided for convenience.
+		 */
+		@Nullable HitResult hitResult();
+
+		/**
+		 * Call to prevent the default block outline from rendering.
+		 * Use this for event handlers that fully replace the vanilla
+		 * block outline render.
+		 *
+		 * <p>The effect of this method depends on the event from which it is called:
+		 *
+		 * <p>In {@code BEFORE_BLOCK_RENDER} this prevents invocation of
+		 * the {@code BLOCK_RENDER} event and {@link AfterBlockOutline#didCancelDefaultBlockOutline()}
+		 * will be {@code false} in later-stage event invocations. It (obviously)
+		 * also prevents the default outline render from running.
+		 *
+		 * <p>In {@code BLOCK_RENDER} this prevents the default outline render
+		 * from running and {@link AfterBlockOutline#didCancelDefaultBlockOutline()}
+		 * will be {@code true} in later-stage event invocations.
+		 *
+		 * <p>In both cases, this method is idempotent - calling it from more
+		 * than one event subscriber has the same effect as calling it once.
+		 * All event handlers within the same event invocation will always
+		 * execute and the cancellation status cannot be inspected during the
+		 * block outline render events.
+		 */
+		void cancelDefaultBlockOutline();
+	}
+
+	/**
+	 * Used in {@code BLOCK_OUTLINE} to convey the parameters normally sent to
+	 * {@code WorldRenderer.drawBlockOutline}.
+	 */
+	public interface BlockOutline {
+		VertexConsumer vertexConsumer();
+
+		Entity entity();
+
+		double cameraX();
+
+		double cameraY();
+
+		double cameraZ();
+
+		BlockPos blockPos();
+
+		BlockState blockState();
+	}
+
+	/**
+	 * Used after the block outline events to test if {@link #cancelDefaultBlockOutline()}
+	 * was called during the {@code BLOCK_OUTLINE} event.
+	 */
+	public interface AfterBlockOutline {
+		/**
+		 * Tests if the default block outline render was cancelled in the {@code BLOCK_OUTLINE} event.
+		 *
+		 * <p>Will NOT be true if the outline was cancelled during {@code BEFORE_BLOCK_OUTLINE}.
+		 *
+		 * @return {@code true} if the default block outline render was cancelled in the {@code BLOCK_OUTLINE} event.
+		 */
+		boolean didCancelDefaultBlockOutline();
+	}
+
+	public interface FrustumContext extends WorldRenderContext, AfterFrustum { }
```

Comment:
Considering the `FrustrumContext` is only used twice, it may be neater to just specify the method across both interface impls. May want to consider this as to reduce the amount of interfaces sub implementing eachother.

Reply:
Not following what "both" refers to.  FrustContext is the parent type for every context other than the starting one.

![image](https://user-images.githubusercontent.com/6106235/100525101-ae958780-3172-11eb-82e3-f25144637a68.png)

Reply:
IMO some nullable getters wouldn't hurt, e.g. adding nullable `frustum()` to `WorldRenderContext` allows dropping both `AfterFrustum` and `FrustumContext`. Nullable `consumers` in `WorldRenderContext` further removes `Drawable` and `MainContext`. `AfterBlockOutline` can IMO also be in `WorldRenderContext` without much potential for confusion. `BlockOutline` could be passed to the appropriate event as a separate parameter (even if it has the same backing object).

With all of the above changes we'd have a single context type everywhere, `WorldRenderContext`, at the cost of a few nullable methods that only yield a proper result in some documented scenarios.

`Drawable` (actually `VertexConsumerProvider` then) may be nicer to pass in the event signature similar to what i proposed for `BlockOutline` above.

Reply:
I'm fine with using `@Nullable` instead of separate instances if that is preferable.

I don't like varying the method signature because I think what may be needed by consumers of these events is still likely to change over time and having everything in the context avoids the need for breaking changes and/or hanging on to deprecated event signatures.  The context is also a more robust extension point for Canvas or others that want to go there.

Reply:
I am not talking about varying the signature everywhere, but passing the big differentiators (BlockOutline and VertexConsumerProvider) directly where it is appropriate, next to the regular context that captures all the generic info.

Reply:
VertexConsumerProvider is specifically on I'd like to keep out of the method signature to better support extension use cases where it (along with the context itself) may be sub-typed. I'd prefer to make it nullable.